File: src/crack/state.py
from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime
from pathlib import Path
import re


def find_repo_root(start: Path | None = None) -> Path:
    current = (start or Path.cwd()).resolve()
    for path in (current, *current.parents):
        if (path / ".git").exists():
            return path
    return current


def timestamp() -> str:
    return datetime.now().astimezone().strftime("%Y-%m-%d %H:%M")


def quote_prompt(prompt: str) -> str:
    lines = prompt.strip().splitlines() or [""]
    return "\n".join(f"> {line}" if line else ">" for line in lines)


def slugify(value: str, default: str = "request") -> str:
    slug = re.sub(r"[^A-Za-z0-9._-]+", "-", value.strip()).strip("-")
    return slug or default


def branch_to_plan_dir(branch_name: str) -> str:
    return slugify(branch_name)


def title_from_prompt(prompt: str) -> str:
    first_line = next((line.strip() for line in prompt.splitlines() if line.strip()), "")
    if not first_line:
        return "User Request"
    return first_line[:80]


@dataclass(frozen=True)
class PlanPaths:
    directory: Path
    plan: Path
    queue: Path
    log: Path


@dataclass(frozen=True)
class ActivePlan:
    paths: PlanPaths
    branch_name: str
    plan_markdown: str
    queue_markdown: str


class MarkdownState:
    def __init__(self, repo_root: Path | str | None = None) -> None:
        self.repo_root = Path(repo_root).resolve() if repo_root else find_repo_root()
        self.crack_dir = self.repo_root / ".crack"
        self.inbox_path = self.crack_dir / "inbox.md"
        self.pr_lock_path = self.crack_dir / "pr-lock.md"
        self.plans_dir = self.crack_dir / "plans"

    def initialize(self) -> None:
        self.plans_dir.mkdir(parents=True, exist_ok=True)
        if not self.inbox_path.exists():
            self.inbox_path.write_text("# Inbox\n\n", encoding="utf-8")

    def read_pr_lock(self) -> str | None:
        if not self.pr_lock_path.exists():
            return None
        return self.pr_lock_path.read_text(encoding="utf-8")

    def plan_paths(self, branch_name: str) -> PlanPaths:
        directory = self.plans_dir / branch_to_plan_dir(branch_name)
        return PlanPaths(
            directory=directory,
            plan=directory / "plan.md",
            queue=directory / "queue.md",
            log=directory / "log.md",
        )

    def existing_plan_paths(self, plan_path: Path | str) -> PlanPaths:
        path = Path(plan_path)
        if not path.is_absolute():
            path = self.repo_root / path
        directory = path if path.is_dir() else path.parent
        return PlanPaths(
            directory=directory,
            plan=directory / "plan.md",
            queue=directory / "queue.md",
            log=directory / "log.md",
        )

    def scan_active_plans(self) -> list[ActivePlan]:
        self.initialize()
        plans: list[ActivePlan] = []
        if not self.plans_dir.exists():
            return plans

        for directory in sorted(path for path in self.plans_dir.iterdir() if path.is_dir()):
            paths = self.existing_plan_paths(directory)
            if not paths.plan.exists():
                continue

            plan_markdown = paths.plan.read_text(encoding="utf-8")
            queue_markdown = paths.queue.read_text(encoding="utf-8") if paths.queue.exists() else ""
            plans.append(
                ActivePlan(
                    paths=paths,
                    branch_name=branch_from_plan(plan_markdown, directory.name),
                    plan_markdown=plan_markdown,
                    queue_markdown=queue_markdown,
                )
            )

        return plans

    def append_inbox(
        self,
        prompt: str,
        reason: str,
        received_at: str | None = None,
    ) -> Path:
        self.initialize()
        entry = (
            "## Queued Request\n\n"
            f"Received: {received_at or timestamp()}\n\n"
            "User prompt:\n\n"
            f"{quote_prompt(prompt)}\n\n"
            "Reason:\n\n"
            f"{reason.strip()}\n\n"
        )
        with self.inbox_path.open("a", encoding="utf-8") as file:
            file.write(entry)
        return self.inbox_path

    def create_plan(
        self,
        branch_name: str,
        plan_title: str,
        prompt: str,
        reason: str,
        received_at: str | None = None,
    ) -> PlanPaths:
        self.initialize()
        branch = branch_name.strip()
        if not branch:
            raise ValueError("branch_name is required")

        title = plan_title.strip() or title_from_prompt(prompt)
        paths = self.plan_paths(branch)
        if paths.directory.exists():
            raise FileExistsError(f"Plan already exists: {paths.directory}")

        paths.directory.mkdir(parents=True)
        received = received_at or timestamp()
        paths.plan.write_text(
            (
                f"# Plan: {title}\n\n"
                f"Branch: {branch}\n\n"
                "## Intent\n\n"
                f"{prompt.strip()}\n\n"
                "## Commit Units\n\n"
                "- TODO: Planner should split this request into commit-sized units.\n\n"
                "## Router Note\n\n"
                f"{reason.strip()}\n"
            ),
            encoding="utf-8",
        )
        paths.queue.write_text("# Queue\n\n", encoding="utf-8")
        paths.log.write_text(
            (
                "# Log\n\n"
                f"## {received}\n\n"
                f"- Created plan directory for `{branch}`.\n"
                f"- Reason: {reason.strip()}\n"
            ),
            encoding="utf-8",
        )
        return paths

    def append_queue(
        self,
        plan_path: Path | str,
        prompt: str,
        reason: str,
        received_at: str | None = None,
    ) -> Path:
        self.initialize()
        paths = self.existing_plan_paths(plan_path)
        if not paths.directory.exists():
            raise FileNotFoundError(f"Plan does not exist: {paths.directory}")
        if not paths.queue.exists():
            paths.queue.write_text("# Queue\n\n", encoding="utf-8")

        entry = (
            "## Queued Request\n\n"
            f"Received: {received_at or timestamp()}\n\n"
            "User prompt:\n\n"
            f"{quote_prompt(prompt)}\n\n"
            "Reason:\n\n"
            f"{reason.strip()}\n\n"
        )
        with paths.queue.open("a", encoding="utf-8") as file:
            file.write(entry)
        return paths.queue

    def set_pr_lock(
        self,
        pr_url: str,
        branch_name: str,
        reason: str,
        status: str = "reviewing",
    ) -> Path:
        self.initialize()
        branch = branch_name.strip()
        if not branch:
            raise ValueError("branch_name is required")

        self.pr_lock_path.write_text(
            (
                "# PR Lock\n\n"
                f"Branch: {branch}\n"
                f"PR: {pr_url.strip()}\n"
                f"Status: {status.strip()}\n\n"
                f"{reason.strip()}\n"
            ),
            encoding="utf-8",
        )
        return self.pr_lock_path

    def clear_pr_lock(self) -> bool:
        if not self.pr_lock_path.exists():
            return False
        self.pr_lock_path.unlink()
        return True


def branch_from_plan(markdown: str, fallback: str) -> str:
    match = re.search(r"^Branch:\s*(.+?)\s*$", markdown, flags=re.MULTILINE)
    if match:
        return match.group(1).strip()
    return fallback

File: src/state.ts
import { existsSync } from "node:fs";
import { appendFile, mkdir, readdir, readFile, unlink, writeFile } from "node:fs/promises";
import path from "node:path";

import { summarizePlanStatus } from "./plan-status";
import type { PlanStatus, PlanStatusSummary } from "./plan-status";

export type PlanPaths = {
  directory: string;
  plan: string;
  queue: string;
  log: string;
};

export type ActivePlan = PlanPaths & {
  branchName: string;
  planContent: string;
  queueContent: string;
};

export type PlanRecord = ActivePlan & {
  logContent: string;
  status: PlanStatus;
  statusSummary: PlanStatusSummary;
};

export type QueuedRequest = {
  prompt: string;
  reason: string;
  receivedAt: string;
};

export function findRepoRoot(start = process.cwd()): string {
  let current = path.resolve(start);

  while (true) {
    if (existsSync(path.join(current, ".git"))) {
      return current;
    }

    const parent = path.dirname(current);
    if (parent === current) {
      return path.resolve(start);
    }

    current = parent;
  }
}

export function timestamp(date = new Date()): string {
  const pad = (value: number) => value.toString().padStart(2, "0");

  return [
    date.getFullYear(),
    pad(date.getMonth() + 1),
    pad(date.getDate()),
  ].join("-")
    + " "
    + [pad(date.getHours()), pad(date.getMinutes())].join(":");
}

export function quotePrompt(prompt: string): string {
  const lines = prompt.trim().split(/\r?\n/);
  const nonEmptyLines = lines.length > 0 ? lines : [""];

  return nonEmptyLines.map((line) => (line ? `> ${line}` : ">")).join("\n");
}

export function slugify(value: string, fallback = "request"): string {
  const slug = value
    .trim()
    .replace(/[^A-Za-z0-9._-]+/g, "-")
    .replace(/\.{2,}/g, ".")
    .replace(/^[.-]+|[.-]+$/g, "")
    .replace(/\.lock$/i, "-lock");

  return slug || fallback;
}

export function planDirectoryName(branchName: string): string {
  return slugify(branchName);
}

export function titleFromPrompt(prompt: string): string {
  const firstLine = prompt.split(/\r?\n/).find((line) => line.trim().length > 0);
  return firstLine ? firstLine.trim().slice(0, 80) : "User Request";
}

export class MarkdownState {
  readonly repoRoot: string;
  readonly crackDir: string;
  readonly inboxPath: string;
  readonly prLockPath: string;
  readonly plansDir: string;

  constructor(repoRoot = findRepoRoot()) {
    this.repoRoot = path.resolve(repoRoot);
    this.crackDir = path.join(this.repoRoot, ".crack");
    this.inboxPath = path.join(this.crackDir, "inbox.md");
    this.prLockPath = path.join(this.crackDir, "pr-lock.md");
    this.plansDir = path.join(this.crackDir, "plans");
  }

  async initialize(): Promise<void> {
    await mkdir(this.plansDir, { recursive: true });

    if (!existsSync(this.inboxPath)) {
      await writeFile(this.inboxPath, "# Inbox\n\n", "utf8");
    }
  }

  async readPrLock(): Promise<string | null> {
    if (!existsSync(this.prLockPath)) {
      return null;
    }

    return readFile(this.prLockPath, "utf8");
  }

  planPaths(branchName: string): PlanPaths {
    const directory = path.join(this.plansDir, planDirectoryName(branchName));

    return {
      directory,
      plan: path.join(directory, "plan.md"),
      queue: path.join(directory, "queue.md"),
      log: path.join(directory, "log.md"),
    };
  }

  existingPlanPaths(planPath: string): PlanPaths {
    const absolutePath = path.isAbsolute(planPath) ? planPath : path.join(this.repoRoot, planPath);
    const directory = path.basename(absolutePath) === "plan.md" ? path.dirname(absolutePath) : absolutePath;

    return {
      directory,
      plan: path.join(directory, "plan.md"),
      queue: path.join(directory, "queue.md"),
      log: path.join(directory, "log.md"),
    };
  }

  async listPlanRecords(options: { initialize?: boolean } = {}): Promise<PlanRecord[]> {
    if (options.initialize ?? true) {
      await this.initialize();
    } else if (!existsSync(this.plansDir)) {
      return [];
    }

    const entries = await readdir(this.plansDir, { withFileTypes: true });
    const plans: PlanRecord[] = [];

    for (const entry of entries) {
      if (!entry.isDirectory()) {
        continue;
      }

      const directory = path.join(this.plansDir, entry.name);
      const plan = path.join(directory, "plan.md");

      if (!existsSync(plan)) {
        continue;
      }

      const queue = path.join(directory, "queue.md");
      const log = path.join(directory, "log.md");
      const planContent = await readFile(plan, "utf8");
      const queueContent = existsSync(queue) ? await readFile(queue, "utf8") : "";
      const logContent = existsSync(log) ? await readFile(log, "utf8") : "";
      const statusSummary = summarizePlanStatus(planContent, logContent);

      plans.push({
        directory,
        plan,
        queue,
        log,
        branchName: branchNameFromPlan(planContent) ?? entry.name,
        planContent,
        queueContent,
        logContent,
        status: statusSummary.status,
        statusSummary,
      });
    }

    return plans.sort((left, right) => left.directory.localeCompare(right.directory));
  }

  async listRoutablePlans(options: { initialize?: boolean } = {}): Promise<PlanRecord[]> {
    const plans = await this.listPlanRecords(options);
    return plans.filter((plan) => plan.statusSummary.routing.routeToExistingPlanCandidate);
  }

  /** @deprecated Use listRoutablePlans for routing candidates or listPlanRecords for all plans. */
  async listActivePlans(): Promise<ActivePlan[]> {
    return this.listRoutablePlans();
  }

  async appendInbox(prompt: string, reason: string, receivedAt = timestamp()): Promise<string> {
    await this.initialize();
    await appendFile(this.inboxPath, queuedRequestEntry(prompt, reason, receivedAt), "utf8");

    return this.inboxPath;
  }

  async readInboxRequests(): Promise<QueuedRequest[]> {
    await this.initialize();
    const content = existsSync(this.inboxPath) ? await readFile(this.inboxPath, "utf8") : "";

    return parseQueuedRequests(content);
  }

  async writeInboxRequests(requests: QueuedRequest[]): Promise<string> {
    await this.initialize();
    await writeFile(this.inboxPath, inboxTemplate(requests), "utf8");

    return this.inboxPath;
  }

  async createPlan(options: {
    branchName: string;
    planTitle: string;
    prompt: string;
    reason: string;
    receivedAt?: string;
  }): Promise<PlanPaths> {
    await this.initialize();

    const branchName = options.branchName.trim();
    if (!branchName) {
      throw new Error("branchName is required");
    }

    const planTitle = options.planTitle.trim() || titleFromPrompt(options.prompt);
    const paths = this.planPaths(branchName);

    if (existsSync(paths.directory)) {
      throw new Error(`Plan already exists: ${paths.directory}`);
    }

    await mkdir(paths.directory, { recursive: true });
    await writeFile(paths.plan, planTemplate(planTitle, branchName, options.prompt, options.reason), "utf8");
    await writeFile(paths.queue, "# Queue\n\n", "utf8");
    await writeFile(paths.log, logTemplate(branchName, options.reason, options.receivedAt ?? timestamp()), "utf8");

    return paths;
  }

  async appendQueue(planPath: string, prompt: string, reason: string, receivedAt = timestamp()): Promise<string> {
    await this.initialize();

    const paths = this.existingPlanPaths(planPath);
    if (!existsSync(paths.directory)) {
      throw new Error(`Plan does not exist: ${paths.directory}`);
    }

    if (!existsSync(paths.queue)) {
      await writeFile(paths.queue, "# Queue\n\n", "utf8");
    }

    await appendFile(paths.queue, queuedRequestEntry(prompt, reason, receivedAt), "utf8");

    return paths.queue;
  }

  async appendPlanLog(planPath: string | PlanPaths, entries: string[], loggedAt = timestamp()): Promise<string> {
    await this.initialize();

    const paths = typeof planPath === "string" ? this.existingPlanPaths(planPath) : planPath;
    if (!existsSync(paths.directory)) {
      throw new Error(`Plan does not exist: ${paths.directory}`);
    }

    if (!existsSync(paths.log)) {
      await writeFile(paths.log, "# Log\n\n", "utf8");
    }

    await appendFile(
      paths.log,
      [
        `## ${loggedAt}`,
        "",
        ...entries.map((entry) => `- ${entry}`),
        "",
      ].join("\n"),
      "utf8",
    );

    return paths.log;
  }

  async setPrLock(options: {
    branchName: string;
    prUrl: string;
    reason: string;
    status?: string;
  }): Promise<string> {
    await this.initialize();

    const branchName = options.branchName.trim();
    if (!branchName) {
      throw new Error("branchName is required");
    }

    const content = [
      "# PR Lock",
      "",
      `Branch: ${branchName}`,
      `PR: ${options.prUrl.trim()}`,
      `Status: ${(options.status ?? "reviewing").trim()}`,
      "",
      options.reason.trim(),
      "",
    ].join("\n");

    await writeFile(this.prLockPath, content, "utf8");
    return this.prLockPath;
  }

  async clearPrLock(): Promise<boolean> {
    if (!existsSync(this.prLockPath)) {
      return false;
    }

    await unlink(this.prLockPath);
    return true;
  }
}

function queuedRequestEntry(prompt: string, reason: string, receivedAt: string): string {
  return [
    "## Queued Request",
    "",
    `Received: ${receivedAt}`,
    "",
    "User prompt:",
    "",
    quotePrompt(prompt),
    "",
    "Reason:",
    "",
    reason.trim(),
    "",
  ].join("\n");
}

export function parseQueuedRequests(content: string): QueuedRequest[] {
  return content
    .split(/^## Queued Request\s*$/m)
    .slice(1)
    .map(parseQueuedRequestSection)
    .filter((request): request is QueuedRequest => request !== null);
}

function parseQueuedRequestSection(section: string): QueuedRequest | null {
  const lines = section.split(/\r?\n/);
  const receivedAt = matchLine(section, /^Received:\s*(.+)\s*$/m) ?? "";
  const promptLabelIndex = lines.findIndex((line) => line.trim() === "User prompt:");
  const reasonLabelIndex = lines.findIndex((line) => line.trim() === "Reason:");

  if (promptLabelIndex < 0 || reasonLabelIndex < 0 || reasonLabelIndex <= promptLabelIndex) {
    return null;
  }

  const prompt = trimOuterEmptyLines(lines.slice(promptLabelIndex + 1, reasonLabelIndex))
    .map((line) => {
      if (line === ">") {
        return "";
      }

      return line.startsWith("> ") ? line.slice(2) : line;
    })
    .join("\n")
    .trim();
  const reason = trimOuterEmptyLines(lines.slice(reasonLabelIndex + 1)).join("\n").trim();

  if (!prompt) {
    return null;
  }

  return { prompt, reason, receivedAt };
}

function inboxTemplate(requests: QueuedRequest[]): string {
  if (requests.length === 0) {
    return "# Inbox\n\n";
  }

  return `# Inbox\n\n${requests
    .map((request) => queuedRequestEntry(request.prompt, request.reason, request.receivedAt))
    .join("\n")}`;
}

function matchLine(content: string, pattern: RegExp): string | undefined {
  return content.match(pattern)?.[1]?.trim();
}

function trimOuterEmptyLines(lines: string[]): string[] {
  let start = 0;
  let end = lines.length;

  while (start < end && lines[start].trim() === "") {
    start += 1;
  }

  while (end > start && lines[end - 1].trim() === "") {
    end -= 1;
  }

  return lines.slice(start, end);
}

function planTemplate(title: string, branchName: string, prompt: string, reason: string): string {
  return [
    `# Plan: ${title}`,
    "",
    `Branch: ${branchName}`,
    "",
    "## Intent",
    "",
    prompt.trim(),
    "",
    "## Commit Units",
    "",
    "- TODO: Planner should split this request into commit-sized units.",
    "",
    "## Router Note",
    "",
    reason.trim(),
    "",
  ].join("\n");
}

function logTemplate(branchName: string, reason: string, receivedAt: string): string {
  return [
    "# Log",
    "",
    `## ${receivedAt}`,
    "",
    `- Created plan directory for \`${branchName}\`.`,
    `- Reason: ${reason.trim()}`,
    "",
  ].join("\n");
}

function branchNameFromPlan(content: string): string | undefined {
  const match = content.match(/^Branch:\s*(.+)\s*$/m);
  return match?.[1]?.trim() || undefined;
}
